### PR TITLE
Fix RunUntil and stop leaking watch channel on etcd error

### PR DESCRIPTION
**Pull request by bprashanth**
Addresses https://github.com/GoogleCloudPlatform/kubernetes/issues/6488, and fixes reflector.RunUntil to actually pipe the stop channel down to the watch handler.

### CODE REVIEW

File: pkg/client/cache/reflector.go [hunk 1]
```diff
@@ -149,9 +156,9 @@ func (r *Reflector) listAndWatch() {
 			}
 			return
 		}
-		if err := r.watchHandler(w, &resourceVersion, exitWatch); err != nil {
+		if err := r.watchHandler(w, &resourceVersion, exitWatch, stopCh); err != nil {
```
Comment (lavalamp) [inline pkg/client/cache/reflector.go]: suggest renaming "exitWatch" to "resyncCh" in this function, too.

File: pkg/client/cache/reflector.go [hunk 2]
```diff
@@ -85,21 +85,28 @@ func NewReflector(lw ListerWatcher, expectedType interface{}, store Store, resyn
...
-	go util.Until(func() { r.listAndWatch() }, r.period, stopCh)
+	go util.Until(func() { r.listAndWatch(stopCh) }, r.period, stopCh)
 }
 
 var (
 	// nothing will ever be sent down this channel
 	neverExitWatch <-chan time.Time = make(chan time.Time)
 
+	// a dummy stop channel that never sends a stop signal
+	neverStopWatch <-chan struct{} = make(chan struct{})
```
Comment (lavalamp) [inline pkg/client/cache/reflector.go]: Reuse this one: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/util/util.go#L97

**Comment by lavalamp** [comment]
This looks great, thank you very much! just two tiny nits.

**Comment by bprashanth** [comment]
PTAL

**Comment by lavalamp** [comment]
LGTM, will merge on green

**Comment by bprashanth** [comment]
Shippable looks like https://github.com/GoogleCloudPlatform/kubernetes/issues/5628

**Pull request merged**